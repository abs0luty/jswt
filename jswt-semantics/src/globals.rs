use jswt_ast::high_level::*;
use jswt_common::{SemanticSymbolTable, Symbol, Type};

use crate::convert::Convert;
use crate::error::SemanticError;

pub struct GlobalResolver {
    pub symbols: SemanticSymbolTable,
    pub errors: Vec<SemanticError>,
}

impl Default for GlobalResolver {
    fn default() -> Self {
        Self {
            symbols: SemanticSymbolTable::default(),
            errors: vec![],
        }
    }
}

impl GlobalResolver {
    pub fn resolve(&mut self, ast: &Ast) {
        self.visit_program(&ast.program);
    }
}

impl StatementVisitor for GlobalResolver {
    fn visit_program(&mut self, node: &Program) {
        self.symbols.push_global_scope();
        self.visit_source_elements(&node.source_elements);
        // Don't pop it because the symbol table here is going to be
        // used in the next resolution pass
    }

    fn visit_source_elements(&mut self, node: &SourceElements) {
        for element in &node.source_elements {
            self.visit_source_element(element);
        }
    }

    fn visit_source_element(&mut self, node: &SourceElement) {
        match node {
            SourceElement::FunctionDeclaration(function) => {
                self.visit_function_declaration(function)
            }
            SourceElement::Statement(statement) => self.visit_statement_element(statement),
        }
    }

    fn visit_statement_element(&mut self, node: &StatementElement) {
        match node {
            StatementElement::Block(stmt) => self.visit_block_statement(stmt),
            StatementElement::Empty(_) => {}
            StatementElement::Return(_) => {}
            StatementElement::Variable(stmt) => self.visit_variable_statement(stmt),
            StatementElement::Expression(_) => {}
            StatementElement::If(_) => {}
            StatementElement::Iteration(_) => {}
        };
    }

    fn visit_block_statement(&mut self, _: &BlockStatement) {
        unreachable!()
    }

    fn visit_empty_statement(&mut self, _: &EmptyStatement) {
        unreachable!()
    }

    fn visit_if_statement(&mut self, _: &IfStatement) {
        unreachable!()
    }

    fn visit_iteration_statement(&mut self, _: &IterationStatement) {
        unreachable!()
    }

    fn visit_while_iteration_element(&mut self, _: &WhileIterationElement) {
        unreachable!()
    }

    fn visit_return_statement(&mut self, _: &ReturnStatement) {
        unreachable!()
    }

    fn visit_variable_statement(&mut self, _: &VariableStatement) {
        // add to global symbol table
    }

    fn visit_expression_statement(&mut self, _: &ExpressionStatement) {
        unreachable!()
    }

    fn visit_statement_list(&mut self, node: &StatementList) {
        for statement in &node.statements {
            self.visit_statement_element(statement);
        }
    }

    fn visit_function_declaration(&mut self, node: &FunctionDeclarationElement) {
        // add to global symbol table
        let ident = &node.ident;
        let name = ident.value;
        if self.symbols.lookup_current(&name.into()).is_some() {
            let error = SemanticError::FunctionAlreadyDefined {
                name,
                span: ident.span.to_owned(),
            };
            self.errors.push(error);
        }

        // Determine return type
        let returns = node
            .returns
            .as_ref()
            .map(Type::convert)
            .unwrap_or(Type::Void);

        let parameters = node
            .params
            .parameters
            .iter()
            .map(|param| Type::convert(&param.type_annotation))
            .collect();

        self.symbols.define(
            name.into(),
            Symbol::new(Type::Function(parameters, Box::new(returns)), name),
        );
    }

    fn visit_function_body(&mut self, _: &FunctionBody) {
        unreachable!()
    }
}
