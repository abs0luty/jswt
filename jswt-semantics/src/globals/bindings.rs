use crate::{bindings::*, SemanticError};

use jswt_ast::*;

#[derive(Debug)]
pub struct GlobalBindingsResolver<'a> {
    pub stack: Vec<ClassBinding>,
    pub bindings: &'a mut BindingsTable,
    pub errors: Vec<SemanticError>,
}

impl<'a> GlobalBindingsResolver<'a> {
    pub fn new(bindings: &'a mut BindingsTable) -> Self {
        Self {
            stack: vec![],
            bindings,
            errors: vec![],
        }
    }

    pub fn enter_class_declaration(&mut self, _: &ClassDeclarationElement) {
        let binding = ClassBinding {};
        self.stack.push(binding);
    }

    pub fn exit_class_declaration(&mut self, node: &ClassDeclarationElement) {
        let current = self.stack.pop().unwrap();
        let name = &node.ident.value;
        self.bindings.define(name.clone(), current);
    }
}
