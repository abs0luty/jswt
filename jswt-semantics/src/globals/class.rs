use super::GlobalSemanticResolver;
use crate::SemanticError;
use jswt_ast::{
    visit::{self, Visitor},
    ClassDeclarationElement,
};
use jswt_common::Type;
use jswt_symbols::{BindingsTable, ClassBinding, FunctionSignature, Method};

pub struct ClassDeclarationGlobalContext<'a> {
    class_binding: ClassBinding,
    bindings: &'a mut BindingsTable,
    errors: &'a mut Vec<SemanticError>,
}

impl<'a> ClassDeclarationGlobalContext<'a> {
    /// We need the class reference here to
    /// figure out the class name
    pub fn new(resolver: &'a mut GlobalSemanticResolver, node: &ClassDeclarationElement) -> Self {
        Self {
            class_binding: ClassBinding {
                name: node.ident.value.clone(),
                fields: vec![],
                methods: vec![],
            },
            bindings: resolver.bindings,
            errors: &mut resolver.errors,
        }
    }
}

impl<'a> Visitor for ClassDeclarationGlobalContext<'a> {
    fn visit_class_declaration(&mut self, node: &ClassDeclarationElement) {
        let class_name = node.ident.value.clone();
        // Walk the rest of the class tree
        // Resolve fields, and methods
        visit::walk_class_declaration(self, &node);

        // Add the class binding to the bindings table
        self.bindings
            .insert(class_name, self.class_binding.to_owned());
    }

    fn visit_class_field_declaration(&mut self, node: &jswt_ast::ClassFieldElement) {
        let field_name = node.ident.value.clone();

        // Check if the field already exists
        if self
            .class_binding
            .fields
            .iter()
            .any(|f| f.name == field_name)
        {
            let error = SemanticError::FieldAlreadyDefined {
                name: field_name.clone(),
                span: node.ident.span.to_owned(),
            };
            self.errors.push(error);
        }

        self.class_binding.fields.push(jswt_symbols::Field {
            name: field_name,
            // Fields are aligned in the order they are declared
            index: self.class_binding.fields.len(),
            // We need to get the type of the field
            // from the symbol table to compute the size
            size: 4, // TODO compute type size
            ty: node.type_annotation.ty.clone(),
        });
        visit::walk_class_field_declaration(self, node);
    }

    fn visit_class_method_declaration(&mut self, node: &jswt_ast::ClassMethodElement) {
        let method_name = node.ident.value.clone();

        // Check if the method already exists
        if self
            .class_binding
            .methods
            .iter()
            .any(|m| m.name == method_name)
        {
            let error = SemanticError::MethodAlreadyDefined {
                name: method_name.clone(),
                span: node.ident.span.to_owned(),
            };
            self.errors.push(error);
        }

        // Get the parameters and return type
        let params = node
            .params
            .parameters
            .iter()
            .map(|param| param.type_annotation.ty.clone())
            .collect();
        let returns = node
            .returns
            .as_ref()
            .map(|ty| ty.ty.clone())
            .unwrap_or(Type::Binding("void".into()));

        // Add the method to the class binding
        self.class_binding.methods.push(Method {
            name: method_name,
            signature: FunctionSignature { params, returns },
        });
    }
}

#[cfg(test)]
mod test {
    use crate::SymbolTable;

    use super::*;
    use jswt_assert::assert_debug_snapshot;
    use jswt_parser::Parser;
    use jswt_symbols::BindingsTable;
    use jswt_tokenizer::Tokenizer;

    #[test]
    fn test_context_resolves_class() {
        let mut tokenizer = Tokenizer::default();
        tokenizer.enqueue_source_str(
            "test_context_resolves_class",
            r"
            class Foo {
                bar: i32;
                baz(a: i32): i32 {
                    return 0;
                }
                empty() {}
            }
        ",
        );
        let ast = Parser::new(&mut tokenizer).parse();
        let mut symbols = SymbolTable::default();
        let mut bindings = BindingsTable::default();
        let mut resolver = GlobalSemanticResolver::new(&mut bindings, &mut symbols);
        resolver.resolve(&ast);

        assert_debug_snapshot!(resolver);
    }

    #[test]
    fn test_context_reports_field_already_defined_error() {
        let mut tokenizer = Tokenizer::default();
        tokenizer.enqueue_source_str(
            "test_context_reports_field_already_defined_error",
            r"
            class Foo {
                bar: i32;
                bar: i32;
            }
        ",
        );
        let ast = Parser::new(&mut tokenizer).parse();
        let mut symbols = SymbolTable::default();
        let mut bindings = BindingsTable::default();
        let mut resolver = GlobalSemanticResolver::new(&mut bindings, &mut symbols);
        resolver.resolve(&ast);

        assert_debug_snapshot!(resolver);
    }
}
