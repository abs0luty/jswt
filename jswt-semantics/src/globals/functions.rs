use super::GlobalSemanticResolver;
use crate::SemanticError;
use jswt_ast::{visit::Visitor, FunctionDeclarationElement};
use jswt_common::Type;
use jswt_symbols::{ScopedSymbolTable, Symbol};

pub struct FunctionDeclarationGlobalContext<'a> {
    symbols: &'a mut ScopedSymbolTable,
    errors: &'a mut Vec<SemanticError>,
}

impl<'a> FunctionDeclarationGlobalContext<'a> {
    pub fn new(resolver: &'a mut GlobalSemanticResolver) -> Self {
        Self {
            symbols: resolver.symbols,
            errors: &mut resolver.errors,
        }
    }
}

impl<'a> Visitor for FunctionDeclarationGlobalContext<'a> {
    fn visit_function_declaration(&mut self, node: &FunctionDeclarationElement) {
        let ident = &node.ident;
        let function_name = &ident.value;

        if self.symbols.lookup_current(function_name).is_some() {
            let error = SemanticError::FunctionAlreadyDefined {
                name: function_name.clone(),
                span: ident.span.to_owned(),
            };
            self.errors.push(error);
        }

        // Determine the return type of the function
        let returns = node
            .returns
            .as_ref()
            .map(|it| it.ty.clone())
            .unwrap_or(Type::Binding("void".into()));

        // Determine the types of the function parameters
        let params = node
            .params
            .parameters
            .iter()
            .map(|param| param.type_annotation.ty.clone())
            .collect();

        // TODO - we're eventually going to want to qualify the full function
        // name here scoped to the current module, but for now we'll just
        // use the function name
        self.symbols
            .define(function_name, Symbol::function(params, returns));
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use jswt_assert::assert_debug_snapshot;
    use jswt_parser::Parser;
    use jswt_symbols::BindingsTable;
    use jswt_tokenizer::Tokenizer;

    #[test]
    fn test_context_resolves_function_bindings() {
        let mut tokenizer = Tokenizer::default();
        tokenizer.enqueue_source_str(
            "test_context_resolves_function_bindings",
            r"
        function test(a: i32, b: i32) {
            let x = 99;
        }
 
        function test2(): Array {

        }
        ",
        );
        let ast = Parser::new(&mut tokenizer).parse();
        let mut symbols = ScopedSymbolTable::default();
        let mut bindings = BindingsTable::default();
        let mut resolver = GlobalSemanticResolver::new(&mut bindings, &mut symbols);
        resolver.resolve(&ast);

        assert_debug_snapshot!(resolver);
    }

    #[test]
    fn test_context_reports_duplicate_function_error() {
        let mut tokenizer = Tokenizer::default();
        tokenizer.enqueue_source_str(
            "test_context_reports_duplicate_function_error",
            r"
        function test2(): Array {

        }
        function test2(): Array {

        }
        ",
        );
        let ast = Parser::new(&mut tokenizer).parse();
        let mut symbols = ScopedSymbolTable::default();
        let mut bindings = BindingsTable::default();
        let mut resolver = GlobalSemanticResolver::new(&mut bindings, &mut symbols);
        resolver.resolve(&ast);

        assert_debug_snapshot!(resolver);
    }
}
