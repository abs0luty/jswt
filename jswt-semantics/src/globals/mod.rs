mod symbols;

use jswt_ast::{visit::Visitor, *};
use jswt_symbols::SymbolTable;

use self::symbols::*;
use crate::SemanticError;

#[derive(Debug)]
pub struct GlobalResolver<'a> {
    symbols: GlobalSymbolsResolver<'a>,
}

impl<'a> GlobalResolver<'a> {
    pub fn new(symbols: &'a mut SymbolTable) -> Self {
        Self {
            symbols: GlobalSymbolsResolver::new(symbols),
        }
    }

    pub fn resolve(&mut self, ast: &Ast) {
        self.visit_program(&ast.program);
    }

    pub fn errors(&mut self) -> Vec<SemanticError> {
        let mut errors = vec![];
        errors.append(&mut self.symbols.errors);
        errors
    }
}

impl<'a> Visitor for GlobalResolver<'a> {
    fn visit_program(&mut self, node: &Program) {
        self.symbols.enter_program(node);
        visit::walk_program(self, node);
        self.symbols.exit_program(node);
    }

    fn visit_function_declaration(&mut self, node: &FunctionDeclarationElement) {
        self.symbols.enter_function_declaration(node);
        visit::walk_function_declaration(self, node);
        self.symbols.exit_function_declaration(node);
    }

    fn visit_class_declaration(&mut self, node: &ClassDeclarationElement) {
        self.symbols.enter_class_declaration(node);
        visit::walk_class_declaration(self, node);
        self.symbols.exit_class_declaration(node);
    }

    fn visit_class_constructor_declaration(&mut self, node: &ClassConstructorElement) {
        self.symbols.enter_constructor_declaration(node);
        visit::walk_class_constructor_declaration(self, node);
    }

    fn visit_class_method_declaration(&mut self, node: &ClassMethodElement) {
        self.symbols.enter_method_declaration(node);
        visit::walk_class_method_declaration(self, node);
    }

    fn visit_class_field_declaration(&mut self, node: &ClassFieldElement) {
        self.symbols.enter_field_declaration(node);
        visit::walk_class_field_declaration(self, node);
    }
}

#[cfg(test)]
mod test {

    use super::*;
    use jswt_assert::assert_debug_snapshot;
    use jswt_parser::Parser;
    use jswt_tokenizer::Tokenizer;

    #[test]
    #[ignore]
    fn test_global_resolver_resolves_class_binding() {
        let mut tokenizer = Tokenizer::default();
        tokenizer.enqueue_source_str(
            "test_global_resolver_resolves_class_binding",
            r"
        let global = 55;

        class Array {
            a: i32;
            b: i32;

            constructor(a: i32, b: f32) {}
            method() {}
        }

        function test() {
            let x = 99;
        }
 
        function doSomething(): Array { }
        ",
        );
        let ast = Parser::new(&mut tokenizer).parse();
        let mut symbols = SymbolTable::default();
        let mut resolver = GlobalResolver::new(&mut symbols);
        resolver.resolve(&ast);
        assert_debug_snapshot!(resolver);
    }

    #[test]
    #[ignore]
    fn test_global_resolver_resolves_function_bindings() {
        let mut tokenizer = Tokenizer::default();
        tokenizer.enqueue_source_str(
            "test_global_resolver_resolves_function_bindings",
            r"
        class Array {

        }

        function test(a: i32, b: i32) {
            let x = 99;
        }
 
        function test2(): Array {

        }
        ",
        );
        let ast = Parser::new(&mut tokenizer).parse();
        let mut symbols = SymbolTable::default();
        let mut resolver = GlobalResolver::new(&mut symbols);
        resolver.resolve(&ast);
        assert_debug_snapshot!(resolver);
    }
}
