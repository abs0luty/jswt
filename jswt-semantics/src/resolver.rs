use crate::error::SemanticError;
use crate::{convert::Convert, types::*};
use std::borrow::Borrow;

use jswt_ast::high_level::*;
use jswt_common::{PrimitiveType, SemanticSymbolTable, Symbol, Type, Typeable};

impl Default for Resolver {
    fn default() -> Self {
        Self {
            symbols: SemanticSymbolTable::default(),
            errors: Default::default(),
        }
    }
}

pub struct Resolver {
    pub symbols: SemanticSymbolTable,
    pub errors: Vec<SemanticError>,
}

impl Resolver {
    pub fn new(symbols: SemanticSymbolTable) -> Self {
        Self {
            symbols,
            errors: Default::default(),
        }
    }

    pub fn resolve(&mut self, ast: &mut Ast) {
        self.visit_program(&mut ast.program);
    }
}

impl MutStatementVisitor for Resolver {
    fn visit_program(&mut self, node: &mut Program) {
        // We expect that the symbol table has the
        // global scope on it
        debug_assert_eq!(self.symbols.depth(), 1);
        self.visit_source_elements(&mut node.source_elements);
        // Local scopes should have popped whatever scopes they created
        debug_assert_eq!(self.symbols.depth(), 1);
        self.symbols.pop_scope();
    }

    fn visit_source_elements(&mut self, node: &mut SourceElements) {
        for element in &mut node.source_elements {
            self.visit_source_element(element);
        }
    }

    fn visit_source_element(&mut self, node: &mut SourceElement) {
        match node {
            SourceElement::FunctionDeclaration(function) => {
                self.visit_function_declaration(function)
            }
            SourceElement::Statement(statement) => self.visit_statement_element(statement),
        }
    }

    fn visit_statement_element(&mut self, node: &mut StatementElement) {
        match node {
            StatementElement::Block(stmt) => self.visit_block_statement(stmt),
            StatementElement::Empty(stmt) => self.visit_empty_statement(stmt),
            StatementElement::Return(stmt) => self.visit_return_statement(stmt),
            StatementElement::Variable(stmt) => self.visit_variable_statement(stmt),
            StatementElement::Expression(stmt) => self.visit_expression_statement(stmt),
            StatementElement::If(stmt) => self.visit_if_statement(stmt),
            StatementElement::Iteration(stmt) => self.visit_iteration_statement(stmt),
        }
    }

    fn visit_block_statement(&mut self, node: &mut BlockStatement) {
        self.symbols.push_scope(node.span(), None);
        self.visit_statement_list(&mut node.statements);
        // TODO - before we pop the scope, determine if
        // there are any types we could not resolve
        self.symbols.pop_scope();
    }

    fn visit_empty_statement(&mut self, _: &mut EmptyStatement) {
        // No-op
    }

    fn visit_if_statement(&mut self, node: &mut IfStatement) {
        self.visit_single_expression(&mut node.condition);

        let condition_ty = &node.condition.defined_type();
        if !condition_ty.is_boolean() {
            self.errors.push(SemanticError::TypeAssignmentError {
                span: node.condition.span(),
                lhs: Type::Primitive(PrimitiveType::Boolean),
                rhs: condition_ty.clone(),
            });
        }
    }

    fn visit_iteration_statement(&mut self, node: &mut IterationStatement) {
        match node {
            IterationStatement::While(elem) => self.visit_while_iteration_element(elem),
        }
    }

    fn visit_while_iteration_element(&mut self, node: &mut WhileIterationElement) {
        self.visit_single_expression(&mut node.expression);

        let condition_ty = &node.expression.defined_type();
        if !condition_ty.is_boolean() {
            self.errors.push(SemanticError::TypeAssignmentError {
                span: node.expression.span(),
                lhs: Type::Primitive(PrimitiveType::Boolean),
                rhs: condition_ty.clone(),
            });
        }
    }

    fn visit_return_statement(&mut self, node: &mut ReturnStatement) {
        self.visit_single_expression(&mut node.expression);

        let value = &node.expression.defined_type();
        let expected = self.symbols.scope_return_type().unwrap_or(&Type::Void);
        if expected != value {
            self.errors.push(SemanticError::TypeMismatchError {
                span: node.expression.span(),
                expected: expected.clone(),
                actual: value.clone(),
            });
        }
    }

    fn visit_variable_statement(&mut self, node: &mut VariableStatement) {
        // Visit right hand side of expression
        self.visit_single_expression(&mut node.expression);

        let name = match &node.target {
            AssignableElement::Identifier(ident) => ident.value,
        };

        if self.symbols.lookup_current(&name.into()).is_some() {
            let error = SemanticError::VariableAlreadyDefined {
                name,
                span: node.target.span(),
            };
            self.errors.push(error);
        }

        // Try to determine defined type from the type annotation
        let lhs = node
            .type_annotation
            .as_ref()
            .map(Type::convert)
            .unwrap_or(Type::Unknown);

        // RHS should have been determined from by the visitor
        let rhs = node.expression.defined_type();
        let res = check_assignment(&lhs, &rhs);

        if res.is_unknown() {
            self.errors.push(SemanticError::TypeMismatchError {
                span: node.expression.span(),
                expected: lhs.clone(),
                actual: rhs.clone(),
            });
        }
        // Define the symbol with the correct type annotation in the symbol table
        self.symbols.define(name.into(), Symbol::new(res, name));
    }

    fn visit_expression_statement(&mut self, node: &mut ExpressionStatement) {
        self.visit_single_expression(&mut node.expression);
    }

    fn visit_statement_list(&mut self, node: &mut StatementList) {
        for statement in &mut node.statements {
            self.visit_statement_element(statement);
        }
    }

    fn visit_function_declaration(&mut self, node: &mut FunctionDeclarationElement) {
        // Push a new local scope for the function body
        self.symbols
            .push_scope(node.span(), node.returns.as_ref().map(Type::convert));
        // Add function parameters as variables in scope
        node.params.parameters.iter().for_each(|param| {
            // Resolve Type from Type Annotation
            let param_name = param.ident.value;
            let ty = Type::convert(&param.type_annotation);
            self.symbols
                .define(param_name.into(), Symbol::new(ty, param_name));
        });

        self.visit_function_body(&mut node.body);
        self.symbols.pop_scope();
    }

    fn visit_function_body(&mut self, node: &mut FunctionBody) {
        self.visit_source_elements(&mut node.source_elements);
    }
}

impl MutExpressionVisitor<()> for Resolver {
    fn visit_single_expression(&mut self, node: &mut SingleExpression) {
        match node {
            SingleExpression::Arguments(exp) => self.visit_argument_expression(exp),
            SingleExpression::Identifier(ident) => self.visit_identifier_expression(ident),
            SingleExpression::MemberIndex(exp) => self.visit_member_index(exp),
            SingleExpression::Assignment(exp) => self.visit_assignment_expression(exp),
            SingleExpression::Multiplicative(exp) => self.visit_binary_expression(exp),
            SingleExpression::Additive(exp) => self.visit_binary_expression(exp),
            SingleExpression::Equality(exp) => self.visit_binary_expression(exp),
            SingleExpression::Bitwise(exp) => self.visit_binary_expression(exp),
            SingleExpression::Relational(exp) => self.visit_binary_expression(exp),
            SingleExpression::Unary(exp) => self.visit_unary_expression(exp),
            SingleExpression::Literal(lit) => self.visit_literal(lit),
        }
    }

    fn visit_assignable_element(&mut self, node: &mut AssignableElement) {
        match node {
            AssignableElement::Identifier(ident) => {
                if let Some(sym) = self.symbols.lookup(ident.value.into()) {}
            }
        }
    }

    fn visit_member_index(&mut self, node: &mut MemberIndexExpression) {
        self.visit_single_expression(&mut *node.index);
        self.visit_single_expression(&mut *node.target);

        let index_type = &node.index.defined_type();
        if !&node.index.defined_type().is_real_number() {
            self.errors.push(SemanticError::TypeMismatchError {
                span: node.span(),
                expected: Type::Primitive(PrimitiveType::I32),
                actual: index_type.clone(),
            });
        }
    }

    fn visit_identifier_expression(&mut self, node: &mut IdentifierExpression) {
        let ident = &node.ident;
        let name = ident.value;

        // Update the node with the type defined in the symbol table if we
        // know what it is
        if let Some(sym) = self.symbols.lookup(name.into()) {
            node.ty = sym.ty.clone();
            return;
        };

        // We haven't encountered this symbol before
        // report error
        self.errors.push(SemanticError::VariableNotDefined {
            name,
            span: ident.span.to_owned(),
        });
    }

    fn visit_argument_expression(&mut self, node: &mut ArgumentsExpression) {
        let expression = node.ident.borrow();
        match expression {
            // Function calls but be followed by an identifier for now
            SingleExpression::Identifier(exp) => {
                // Check that the args are defined in this scope
                for arg in &mut node.arguments.arguments {
                    self.visit_single_expression(arg);
                }

                let name = exp.ident.value;
                if let Some(sym) = self.symbols.lookup(name.into()) {
                    match &sym.ty {
                        Type::Function(_, returns) => {
                            // The type of a function call node is the
                            // type that the function returns
                            node.ty = *returns.clone();
                        }
                        // Someone is trying to invoke something that isn't a function
                        _ => self.errors.push(SemanticError::NotAFunctionError {
                            span: node.span(),
                            name_span: exp.span(),
                        }),
                    }
                    return;
                }

                // We haven't encountered the symbol before
                // report diagnostic error
                self.errors.push(SemanticError::FunctionNotDefined {
                    span: node.span(),
                    name_span: exp.ident.span.to_owned(),
                });
            }

            // We don't have higher order functions yet. if the target of the call
            // is not an identifier report an error
            exp => self.errors.push(SemanticError::NotAFunctionError {
                span: node.span(),
                name_span: exp.span(),
            }),
        };
    }

    fn visit_unary_expression(&mut self, node: &mut UnaryExpression) {
        self.visit_single_expression(&mut node.expr);
        let exp = node.expr.defined_type();

        match node.op {
            UnaryOperator::Plus(_) | UnaryOperator::Minus(_) => {
                if !exp.is_signed_number() {
                    self.errors.push(SemanticError::TypeMismatchError {
                        span: node.expr.span(),
                        // TODO generify this and provide alternatives
                        expected: Type::i32(),
                        actual: exp.clone(),
                    });
                }
                node.ty = exp;
            }
            UnaryOperator::Not(_) => {
                if !exp.is_real_number() {
                    self.errors.push(SemanticError::TypeMismatchError {
                        span: node.expr.span(),
                        expected: Type::i32(),
                        actual: exp.clone(),
                    });
                }
                node.ty = exp;
            }
        }
    }

    fn visit_assignment_expression(&mut self, node: &mut BinaryExpression) {
        self.visit_single_expression(&mut node.left);
        self.visit_single_expression(&mut node.right);

        let lhs = &node.left.defined_type();
        let rhs = &node.right.defined_type();

        // Check that we can perform the assignment
        let res = check_assignment(lhs, rhs);
        if res.is_unknown() {
            self.errors.push(SemanticError::TypeAssignmentError {
                span: node.span(),
                lhs: lhs.clone(),
                rhs: rhs.clone(),
            });
        }

        // If we don't know the type of the LHS update the symbol table
        if lhs.is_unknown() {
            if let SingleExpression::Identifier(exp) = node.left.borrow() {
                self.symbols
                    .update_type(&exp.ident.value.into(), rhs.clone())
            }
        }

        // The node itself is always void because you can't
        // operate on an assignment
        node.ty = Type::Void;
    }

    fn visit_binary_expression(&mut self, node: &mut BinaryExpression) {
        self.visit_single_expression(&mut node.left);
        self.visit_single_expression(&mut node.right);

        let lhs = &node.left.defined_type();
        let rhs = &node.right.defined_type();

        match node.op {
            BinaryOperator::Plus(_)
            | BinaryOperator::Minus(_)
            | BinaryOperator::Mult(_)
            | BinaryOperator::Div(_) => {
                let res = check_arithmetic(&lhs, &rhs);
                if res.is_unknown() {
                    self.errors.push(SemanticError::TypeBinaryOperationError {
                        span: node.span(),
                        op: node.op.clone(),
                        lhs: lhs.clone(),
                        rhs: rhs.clone(),
                    })
                }
                node.ty = res;
            }
            BinaryOperator::Equal(_)
            | BinaryOperator::NotEqual(_)
            | BinaryOperator::Greater(_)
            | BinaryOperator::GreaterEqual(_)
            | BinaryOperator::Less(_)
            | BinaryOperator::LessEqual(_) => {
                let res = check_comparison(&lhs, &rhs);
                if res.is_unknown() {
                    self.errors.push(SemanticError::TypeBinaryOperationError {
                        span: node.span(),
                        op: node.op.clone(),
                        lhs: lhs.clone(),
                        rhs: rhs.clone(),
                    })
                }
                node.ty = res;
            }
            BinaryOperator::And(_) | BinaryOperator::Or(_) => {
                let res = check_bitwise(&lhs, &rhs);
                if res.is_unknown() {
                    self.errors.push(SemanticError::TypeBinaryOperationError {
                        span: node.span(),
                        op: node.op.clone(),
                        lhs: lhs.clone(),
                        rhs: rhs.clone(),
                    });
                }
                node.ty = res;
            }
            BinaryOperator::Assign(_) => {
                let res = check_assignment(&lhs, &rhs);
                if res.is_unknown() {
                    self.errors.push(SemanticError::TypeBinaryOperationError {
                        span: node.span(),
                        op: node.op.clone(),
                        lhs: lhs.clone(),
                        rhs: rhs.clone(),
                    });
                }
                node.ty = Type::Void;
            }
        }
    }

    fn visit_literal(&mut self, literal: &mut Literal) {
        match literal {
            Literal::Array(a) => {
                let mut ty = Type::unknown();
                for e in a.elements.iter_mut() {
                    self.visit_single_expression(e);
                    // if we don't know the type at this point
                    // The type of the first element is our type
                    let defined_type = e.defined_type();
                    if ty.is_unknown() {
                        ty = defined_type;
                    } else if ty != defined_type {
                        ty = Type::unknown();
                        // error
                        break;
                    }
                }
                // Set the type of the array
                a.ty = Type::array(ty);
            }
            _ => {
                // Type literals have their types embedded
            }
        }
    }
}
