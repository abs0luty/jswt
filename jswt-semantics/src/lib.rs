mod convert;
mod error;
mod globals;
mod resolver;
mod types;

use jswt_ast::high_level::Ast;

pub use error::*;
use globals::*;
use jswt_common::SemanticSymbolTable;
use resolver::*;

pub struct SemanticAnalyzer;

pub struct SemanticData {
    pub errors: Vec<SemanticError>,
    pub symbols: SemanticSymbolTable,
}

impl SemanticAnalyzer {
    pub fn analyze(ast: &mut Ast) -> SemanticData {
        let mut global_resolver = GlobalResolver::default();
        global_resolver.resolve(ast);

        let symbols = global_resolver.symbols;
        let mut resolver = Resolver::new(symbols);
        resolver.resolve(ast);

        let mut errors = vec![];
        errors.append(&mut global_resolver.errors);
        errors.append(&mut resolver.errors);

        SemanticData {
            errors,
            symbols: resolver.symbols,
        }
    }
}

#[cfg(test)]
mod test {

    use super::*;
    use jswt_assert::assert_debug_snapshot;
    use jswt_parser::Parser;
    use jswt_tokenizer::Tokenizer;

    #[test]
    fn test_duplicate_variable_declaration_generates_error() {
        let mut tokenizer = Tokenizer::default();
        tokenizer.enqueue_source_str("test.1", "function test() { let x = 0; let x = 1; }");
        let mut ast = Parser::new(&mut tokenizer).parse();
        let data = SemanticAnalyzer::analyze(&mut ast);
        assert_debug_snapshot!(data.errors);
    }

    #[test]
    fn test_variable_not_defined_generates_error() {
        let mut tokenizer = Tokenizer::default();
        tokenizer.enqueue_source_str("test.1", "function test() { return x; }");
        let mut ast = Parser::new(&mut tokenizer).parse();
        let data = SemanticAnalyzer::analyze(&mut ast);
        assert_debug_snapshot!(data.errors);
    }

    #[test]
    fn test_function_return_type_is_checked_when_invoked_error() {
        let mut tokenizer = Tokenizer::default();
        tokenizer.enqueue_source_str(
            "test.1",
            r#"
            function main():i32 { 
                return test(); 
            } 
            
            function test(): boolean { 
                return x; 
            }
        "#,
        );
        let mut ast = Parser::new(&mut tokenizer).parse();
        let data = SemanticAnalyzer::analyze(&mut ast);
        assert_debug_snapshot!(data.errors);
    }

    #[test]
    fn test_if_statement_expression_must_be_boolean() {
        let mut tokenizer = Tokenizer::default();
        tokenizer.enqueue_source_str(
            "test.1",
            r#"
            function main():i32 { 
                if(test()) {

                }
            } 
            
            function test(): i32 { 
                return 0; 
            }
        "#,
        );
        let mut ast = Parser::new(&mut tokenizer).parse();
        let data = SemanticAnalyzer::analyze(&mut ast);
        assert_debug_snapshot!(data.errors);
    }
}
