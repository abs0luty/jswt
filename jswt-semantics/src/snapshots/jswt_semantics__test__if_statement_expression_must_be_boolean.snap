---
source: jswt-semantics/src/lib.rs
expression: errors

---
[
    TypeAssignmentError {
        span: Span {
            file: "test.1",
            start: 55,
            end: 61,
        },
        lhs: Primitive(
            Boolean,
        ),
        rhs: Primitive(
            I32,
        ),
    },
]
